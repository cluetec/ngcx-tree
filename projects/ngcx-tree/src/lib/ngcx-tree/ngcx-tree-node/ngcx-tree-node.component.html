<div
  class="tree-node-content-container"
  [class.selected]="isSelected"
  [class.first]="nodeWrapper.isFirstChild"
  [class.last]="nodeWrapper.isLastChild"
  [class.is-selectable]="nodeWrapper.isSelectable"
  (click)="clickEvent.emit()">
  <div class="tree-node-expand-container">
    <button
      *ngIf="nodeWrapper.children.length > 0"
      class="tree-node-expand"
      cdkTreeNodeToggle
      [attr.aria-label]="'Toggle ' + nodeWrapper.data.title"
      (click)="$event.preventDefault()">
      <div class="chevron" [class.rotate]="treeControl.isExpanded(nodeWrapper)">
        <svg
          id="a"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 9 9"
          height="10px"
          width="10px">
          <path
            d="m2.83.09l4.2,4.2c.05.05.09.14.09.21s-.04.15-.09.21l-4.2,4.2c-.05.05-.14.09-.21.09s-.15-.04-.21-.09l-.45-.45c-.05-.05-.09-.13-.09-.21,0-.07.04-.15.09-.21l3.54-3.54L1.97.96c-.05-.05-.09-.14-.09-.21s.04-.15.09-.21l.45-.45c.05-.05.14-.09.21-.09s.15.04.21.09Z"
            style="fill: var(--icon-color, #333); stroke-width: 0px" />
        </svg>
      </div>
    </button>
  </div>
  <div class="small-horizontal-tree-line"></div>
  <div class="tree-node-icon-container" *ngIf="nodeWrapper.data.faIcon">
    <i class="fa" [class]="nodeWrapper.data.faIcon"></i>
  </div>
  <div class="tree-node-content">
    <span
      *ngIf="
        !treeConfig?.treeNodeContentComponent &&
        !treeConfig?.treeNodeContentTemplate
      ">
      {{ nodeWrapper.data.title }}
    </span>
    <ng-container #ref></ng-container>
    <ng-container *ngIf="treeConfig?.treeNodeContentTemplate as template">
      <ng-container
        *ngTemplateOutlet="
          template;
          context: {
            nodeWrapper: nodeWrapper,
          }
        "></ng-container>
    </ng-container>
  </div>
</div>
