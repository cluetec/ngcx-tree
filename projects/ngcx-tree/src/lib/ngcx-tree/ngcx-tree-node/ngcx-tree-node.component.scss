.tree-node-content-container {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  width: 100%;

  &.is-selectable {
    cursor: pointer;
  }

  .tree-node-expand-container {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    width: 30px;
    height: 30px;
    margin-right: 5px;

    > .tree-node-expand {
      color: #2587be;
      background: none;
      border: none;
      font: inherit;
      cursor: pointer;
      outline: inherit;

      width: 30px;
      height: 30px;
      display: flex;
      justify-content: center;
      align-items: center;

      .chevron {
        width: 10px;
        transition: transform 0.1s ease-in-out;

        &.rotate {
          transform: rotate(90deg);
        }
      }
    }
  }

  .tree-node-icon-container {
    margin-right: 7px;
    color: #2587be;
  }
}
