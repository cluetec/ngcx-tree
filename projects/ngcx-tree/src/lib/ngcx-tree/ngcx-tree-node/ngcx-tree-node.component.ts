import { CdkTreeModule, NestedTreeControl } from '@angular/cdk/tree';
import {
  Component,
  EventEmitter,
  Input,
  OnDestroy,
  OnInit,
  Output,
  ViewChild,
  ViewContainerRef,
} from '@angular/core';
import { TreeConfig, TreeNodeWrapper } from '../ngcx-tree-models';

import { NgIf, NgTemplateOutlet } from '@angular/common';
import { Subject, takeUntil } from 'rxjs';

@Component({
  selector: 'ngcx-tree-node',
  templateUrl: './ngcx-tree-node.component.html',
  styleUrls: ['./ngcx-tree-node.component.scss'],
  standalone: true,
  imports: [CdkTreeModule, NgTemplateOutlet, NgIf],
})
export class NgcxTreeNodeComponent implements OnInit, OnDestroy {
  @Input() nodeWrapper!: TreeNodeWrapper;
  @Input() treeControl!: NestedTreeControl<TreeNodeWrapper>;
  @Input() treeConfig?: TreeConfig;

  @Output() customEvent = new EventEmitter<any>();
  @Output() clickEvent = new EventEmitter<void>();

  @ViewChild('ref', { read: ViewContainerRef, static: true })
  vcRef?: ViewContainerRef;

  ngUnsubscribe = new Subject();

  ngOnInit() {
    if (this.vcRef && this.treeConfig?.treeNodeContentComponent) {
      const nodeComponent = this.vcRef.createComponent(
        this.treeConfig.treeNodeContentComponent
      );
      nodeComponent.instance.nodeWrapper = this.nodeWrapper;

      nodeComponent.instance.customEvent
        ?.pipe(takeUntil(this.ngUnsubscribe))
        .subscribe((value: any) => this.customEvent.emit(value));
    }
  }

  ngOnDestroy(): void {
    this.ngUnsubscribe.next(undefined);
    this.ngUnsubscribe.complete();
  }
}
