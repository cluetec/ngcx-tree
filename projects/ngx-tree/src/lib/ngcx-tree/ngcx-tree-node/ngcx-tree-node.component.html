<div
  class="tree-node-content-container"
  [class.first]="nodeWrapper.isFirstChild"
  [class.last]="nodeWrapper.isLastChild"
  (click)="clickEvent.emit()">
  <div class="tree-node-expand-container">
    <button
      *ngIf="nodeWrapper.children.length > 0"
      class="tree-node-expand"
      cdkTreeNodeToggle
      [attr.aria-label]="'Toggle ' + nodeWrapper.node.title"
      (click)="$event.preventDefault()">
      <fa-icon
        [icon]="
          treeControl.isExpanded(nodeWrapper) ? faChevronDown : faChevronRight
        "
        size="xs"
        [attr.aria-hidden]="false"></fa-icon>
    </button>
  </div>
  <div class="small-horizontal-tree-line"></div>
  <div class="tree-node-icon-container">
    <fa-icon [icon]="faBook" [attr.aria-hidden]="false"></fa-icon>
  </div>
  <div class="tree-node-content">
    <span *ngIf="!treeConfig?.treeNodeContentComponent">{{
      nodeWrapper.node.title
    }}</span>
    <ng-container #ref></ng-container>
  </div>
</div>
