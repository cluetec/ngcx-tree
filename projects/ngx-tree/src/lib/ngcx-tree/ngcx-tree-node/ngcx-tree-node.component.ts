import { CdkTreeModule, NestedTreeControl } from '@angular/cdk/tree';
import { CommonModule } from '@angular/common';
import {
  Component,
  Input,
  OnInit,
  Type,
  ViewChild,
  ViewContainerRef,
} from '@angular/core';
import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';
import { TreeNodeWrapper } from '../../ngcx-tree-models';

import {
  faBook,
  faChevronDown,
  faChevronRight,
} from '@fortawesome/free-solid-svg-icons';

@Component({
  selector: 'ngcx-tree-node',
  templateUrl: './ngcx-tree-node.component.html',
  styleUrls: ['./ngcx-tree-node.component.scss'],
  standalone: true,
  imports: [CdkTreeModule, FontAwesomeModule, CommonModule],
})
export class NgcxTreeNodeComponent implements OnInit {
  @Input() nodeWrapper!: TreeNodeWrapper;
  @Input() treeControl!: NestedTreeControl<TreeNodeWrapper>;
  @Input() treeNodeContentComponent?: Type<any>;

  @ViewChild('ref', { read: ViewContainerRef, static: true })
  vcRef?: ViewContainerRef;

  faBook = faBook;
  faChevronDown = faChevronDown;
  faChevronRight = faChevronRight;

  ngOnInit() {
    if (this.vcRef && this.treeNodeContentComponent) {
      const nodeComponent = this.vcRef.createComponent(
        this.treeNodeContentComponent
      );
      nodeComponent.instance.nodeWrapper = this.nodeWrapper;
    }
  }
}
