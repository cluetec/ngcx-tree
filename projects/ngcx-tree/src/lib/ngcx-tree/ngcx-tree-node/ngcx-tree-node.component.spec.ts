/* tslint:disable:no-unused-variable */
import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { NgcxTreeNodeComponent } from './ngcx-tree-node.component';

describe('NgcxTreeNodeComponent', () => {
  let component: NgcxTreeNodeComponent;
  let fixture: ComponentFixture<NgcxTreeNodeComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [NgcxTreeNodeComponent],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(NgcxTreeNodeComponent);

    component = fixture.componentInstance;
    component.nodeWrapper = {
      children: [],
      depth: 1,
      id: 'a1',
      isFirstChild: true,
      isLastChild: true,
      data: {
        id: 'a1',
        title: 'Apple',
      },
    };
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });
});
