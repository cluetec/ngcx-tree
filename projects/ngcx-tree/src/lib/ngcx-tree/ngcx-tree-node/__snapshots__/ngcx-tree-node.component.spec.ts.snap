// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgcxTreeNodeComponent should create 1`] = `
<ngcx-tree-node
  clickEvent={[Function EventEmitter_]}
  customEvent={[Function EventEmitter_]}
  isSelected="false"
  ngUnsubscribe={[Function Subject]}
  nodeWrapper={[Function Object]}
  vcRef={[Function ViewContainerRef2]}
>
  <div
    class="tree-node-content-container first last"
  >
    <div
      class="tree-node-expand-container"
    >
    </div>
    <div
      class="small-horizontal-tree-line"
    />
    <div
      class="tree-node-content"
    >
      <span>
         Apple 
      </span>
    </div>
  </div>
</ngcx-tree-node>
`;
